File: src/main/java/spring/springserver/domain/chat/entity/ChatRoomParticipant.java
package spring.springserver.domain.chat.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import java.time.Instant;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import spring.springserver.domain.member.entity.Member;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Table(
        name = "chat_room_participant",
        uniqueConstraints = {
                @UniqueConstraint(name = "uk_chat_room_participant_room_member", columnNames = {"room_id", "member_id"})
        },
        indexes = {
                @Index(name = "idx_chat_room_participant_room", columnList = "room_id"),
                @Index(name = "idx_chat_room_participant_member", columnList = "member_id"),
                @Index(name = "idx_chat_room_participant_member_visible", columnList = "member_id, visible")
        }
)
public class ChatRoomParticipant {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "room_id", nullable = false)
    private ChatRoom room;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "member_id", nullable = false)
    private Member member;

    @Column(name = "visible", nullable = false)
    private boolean visible;

    @Column(name = "joined_at", nullable = false, updatable = false)
    private Instant joinedAt;

    @Column(name = "deleted_at")
    private Instant deletedAt;

    public ChatRoomParticipant(ChatRoom room,
                               Member member) {

        this.room = room;
        this.member = member;
        this.visible = true;
        this.joinedAt = Instant.now();
    }

    public void leave(Instant deletedAt) {

        this.visible = false;
        this.deletedAt = deletedAt;
    }

    public void reactivate() {

        this.visible = true;
    }
}

File: src/main/java/spring/springserver/domain/chat/entity/ChatRoomParticipant.kt
package spring.springserver.domain.chat.entity

import jakarta.persistence.*
import spring.springserver.domain.member.entity.Member
import java.time.Instant

@Entity
@Table(
    name = "chat_room_participant",
    uniqueConstraints = [
        UniqueConstraint(name = "uk_chat_room_participant_room_member", columnNames = ["room_id", "member_id"])
    ],
    indexes = [
        Index(name = "idx_chat_room_participant_room", columnList = "room_id"),
        Index(name = "idx_chat_room_participant_member", columnList = "member_id"),
        Index(name = "idx_chat_room_participant_member_visible", columnList = "member_id, visible")
    ]
)
class ChatRoomParticipant(
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "room_id", nullable = false)
    var room: ChatRoom,

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "member_id", nullable = false)
    var member: Member,
) {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private var id: Long? = null

    @Column(name = "visible", nullable = false)
    var visible: Boolean = true

    @Column(name = "joined_at", nullable = false, updatable = false)
    var joinedAt: Instant = Instant.now()

    /**
     * 마지막으로 채팅방을 나간 시각.
     *
     * 단순한 삭제 이력이 아니라 "이 시각 이전 메시지는 이 참여자에게 보이지 않는다"는 워터마크로 쓴다.
     * 다시 입장해도(`rejoin`) 이 값을 지우지 않는다. 지우면 나가기 전 메시지가 전부 다시 보인다.
     */
    @Column(name = "deleted_at")
    var deletedAt: Instant? = null

    fun getId(): Long? = id

    fun leave(
        deletedAt: Instant
    ) {

        visible = false
        this.deletedAt = deletedAt
    }

    /**
     * 채팅방을 목록에 다시 노출시킨다.
     *
     * `deletedAt` 워터마크는 의도적으로 유지한다. 나가기 전 메시지는 계속 가려진 상태로 둔다.
     */
    fun reactivate() {

        visible = true
    }
}
